### ISSUE TITLE
Layer selecton for clear_attachments()

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#1952]
`VkClearRect` has extra fields for the layer range, which we need to support as well.

**Comment by fkaa** [comment]
All the different APIs seem to have different limitations for clearing, where vulkan is the least restrictive.

## dx12 
Supports clearing *n* rects same as Vulkan, but clears the entire RTV/DSV (all layers in the view). AFAICT the dx12 backend already creates a RTV/DSV per layer, so this shouldn't too complicated.

## dx11
Clears the entire RTV/DSV just like DX12, can also create view for each layer for clearing. Doesn't support clearing subregions, so it would need to handle clears that dont affect the entire resource (drawing a rectangle? bit more complicated for DSV).

## gl
A framebuffer attachment in GL can only have one layer, so to clear all layers you can have a framebuffer and assign each layer to an attachment with [`glFramebufferTextureLayer`](https://www.khronos.org/opengl/wiki/GLAPI/glFramebufferTextureLayer) and then clear it. Apparently OpenGL clears respect scissor rectangles, so drawing a rect might not be necessary.

## Metal
Everything related to clearing an attachment (base level, layer count, clear value) must be specified when creating a renderpass, so for each unique layer range there would need to be a renderpass. Unfortunately Metal doesn't seem to support subregion clears (same as dx11) so it would need to handle that case as well.

**Comment by kvark** [comment]
@fkaa thanks for investigation!

> Doesn't support clearing subregions, so it would need to handle clears that dont affect the entire resource (drawing a rectangle? bit more complicated for DSV).

Yes, drawing a rectangle would be a required workaround here, similar to Metal.

> A framebuffer attachment in GL can only have one layer

It's definitely not the case for desktop GL, where you can call `glFramebufferTexture` with a 2DArray/3D texture to get a layered FBO. But one needs to check the exact capabilities required and GL ES support (we should target ES 3.0).

**Comment by fkaa** [comment]
Implemented on Vulkan, but I have some concerns about the rest of the backends. Having a RTV per layer doesn't scale very well (think 100+ layers), so allocating from the RTV pool is probably the best choice for DX12.

Not sure how fast `CreateRenderTargetView` in DX11 is, but it's probably acceptable if we create them on demand and cache them?

For GL I think having a single framebuffer object just for clearing and re-configuring it shouldn't be too bad.

Any suggestions?

**Comment by msiglreith** [comment]
> Having a RTV per layer doesn't scale very well (think 100+ layers), so allocating from the RTV pool is probably the best choice for DX12.

My suggestion would be allocating a rtv cpu descriptor heap and create one rtv per clear rect.

**Comment by fkaa** [comment]
@msiglreith Got it working using one rtv per rect (https://github.com/fkaa/gfx/commit/fbc662344f6b2b8d5e0f25c1840083ebae1ba2e7), but I'm storing the heap inside the command buffer so it runs out of descriptors eventually. Is there a way to reset the heap or am I using it incorrectly?

**Comment by msiglreith** [comment]
@fkaa To clarify my sugeestion:
```rust
fn clear_attachments(clears) {
     let rtv_pool = CpuDescriptorPool::new(clears.len()); // +/- a few more clears..
     for (i, clear) in clears.enumerate() {
          let rtv_handle = rtv_pool.at(i);
          device.CreateRenderTargetView(rtv_handle, ...);
          ...
     }
}
```
RTV pools are non-shader visible, so the pool should be immediately allowed for reuse/destruction.

**Comment by kvark** [comment]
Fixed in #2003
